microsoft/vscode-azurefunctions
# Enable continuous debugging for Python functions

Thanks @mhoeger for bringing this up and suggesting for the --wait-for-client change. As Azure Functions Python also want to enable hot-reload debugging on Azure Functions.

### Background
Testing with `"azureFunctions.stopFuncTaskPostDebug": false` in .vscode/settings.json, everytime when a user is in a debug session, making changes to the function code will result in a timeout issue. This is mainly caused by:
1. When the host restart the Python worker, it actually restart the debugpy program instead, e.g.
```
Starting worker process with FileName:py WorkingDirectory:C:\Users\hazeng\Projects\Functions\PythonVSCodeDebug 
Arguments: c:/Users/hazeng/.vscode/extensions/ms-python.python-2020.11.371526539/pythonFiles/lib/python/debugpy
 --listen 127.0.0.1:9091
 --wait-for-client
 "C:\Users\hazeng\Projects\Artifacts\_\workers\python\3.7/WINDOWS/X64/worker.py"
 --host 127.0.0.1
 --port 23161
 --workerId 7ce43683-d867-48ad-9546-d366e0ba1aa1
 --requestId 107191ad-ab4e-45e6-9383-42c430796b04
 --grpcMaxMessageLength 2147483647
```
2. As the `--wait-for-client` flag generated from [PythonDebugProvider.ts](https://github.com/microsoft/vscode-azurefunctions/blob/main/src/debug/PythonDebugProvider.ts#L42) which will block the Python worker from starting (as the vscode debugger quits on file change). In this case, we want to remove the `--wait-for-client` flag, and let the host restarts the worker.

### Testing
1. Download this branch and hit F5 to enable debug
2. Create a new python function app project
3. Add a configuration in .vscode/settings.json, `azureFunctions.stopFuncTaskPostDebug: false`
4. Hit F5 to start debug the Python worker, the host should now spin up a debugpy server with Azure Functions Python Worker running inside.
5. Change some code in the middle of debugging section, and the host should restart the debugpy server with the latest user code change.
6. Reattach the debugger by pressing F5 again. The host never exits as the debugger should now be attached in the same host process.

cc: @anirudhgarg @vrdmr @mhoeger @ejizba
